RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+        result.insert(result.end(),hf.begin(),hf.end());
+    }
+    storeRecHits(result);
+}
+
+void EGammaPCAHelper::storeRecHits(const reco::CaloCluster & cluster) {
+    theCluster_ = &cluster;
+    storeRecHits(cluster.hitsAndFractions());
+}
+
+void EGammaPCAHelper::storeRecHits(const std::vector<std::pair<DetId, float>> &hf) {
+    std::vector<double> pcavars;
+    pcavars.resize(3,0.);
+    theSpots_.clear();
+    pcaIteration_ = 0;
+
+    sigu_ = 0.;
+    sigv_ = 0.;
+    sigp_ = 0.;
+    sige_ = 0.;
+
+    unsigned hfsize = hf.size();
+    if (debug_)
+    std::cout << "The seed cluster constains " << hfsize << " hits " << std::endl;
+
+    if (hfsize == 0) return;
+
+
+    for (unsigned int j = 0; j < hfsize; j++) {
+        unsigned int layer = recHitTools_->getLayerWithOffset(hf[j].first);
+
+        const DetId rh_detid = hf[j].first;
+        std::map<DetId,const HGCRecHit *>::const_iterator itcheck= hitMap_->find(rh_detid);
+        if (itcheck == hitMap_->end()) {
+            std::cout << " Big problem, unable to find a hit " << rh_detid.rawId() << " " ;
+            std::cout << rh_detid.det() << " " << HGCalDetId(rh_detid) << std::endl;
+            continue;
+        }
+        if (debug_) {
+            std::cout << "DetId " << rh_detid.rawId() << " " << layer << " " <<  itcheck->second->energy() <<std::endl;
+            //        std::cout << " Hit " << itcheck->second << " " << itcheck->second->energy() << std::endl;
+        }
+        float fraction = hf[j].second;
+
+        double thickness =
+            (DetId::Forward == DetId(rh_detid).det()) ? recHitTools_->getSiThickness(rh_detid) : -1;
+        double mip = dEdXWeights_[layer] * 0.001;  // convert in GeV
+        if (thickness > 99. && thickness < 101)
+            mip *= invThicknessCorrection_[0];
+        else if (thickness > 199 && thickness < 201)
+            mip *= invThicknessCorrection_[1];
+        else if (thickness > 299 && thickness < 301)
+            mip *= invThicknessCorrection_[2];
+
+        pcavars[0] = recHitTools_->getPosition(rh_detid).x();
+        pcavars[1] = recHitTools_->getPosition(rh_detid).y();
+        pcavars[2] = recHitTools_->getPosition(rh_detid).z();
+        if (pcavars[2] == 0.)
+            std::cout << " Problem, hit with z =0 ";
+        else  {
+            Spot mySpot(rh_detid,itcheck->second->energy(),pcavars,layer,fraction,mip);
+            theSpots_.push_back(mySpot);
+        }
+    }
+    if (debug_) {
+        std::cout << " Stored " << theSpots_.size() << " hits " << std::endl;
+    }
+}
+
+void EGammaPCAHelper::computePCA(float radius , bool withHalo) {
+    // very important - to reset
+    pca_.reset(new TPrincipal(3, "D"));
+    bool initialCalculation = radius < 0;
+    if (debug_)
+        std::cout << " Initial calculation " << initialCalculation << std::endl;
+    if (initialCalculation && withHalo) {
+        std::cout << "Warning - in the first iteration, the halo hits are excluded " << std::endl;
```

Comment:
Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.